### PR TITLE
PSA Crypto Configuration Split

### PR DESCRIPTION
**Pull request by minosgalanakis**
## Description

This PR is migrating the configuration split entries as outlined in the [design document](https://github.com/Mbed-TLS/mbedtls/blob/development/docs/proposed/config-split.md). It also updates the testing framework, to support those changes.

## PR checklist

Please remove the segment/s on either side of the | symbol as appropriate, and add any relevant link/s to the end of the line.
If the provided content is part of the present PR remove the # symbol.

- [x] **changelog** provided
- [x] **development PR** provided # | This is it.
- [x] **framework PR** provided https://github.com/Mbed-TLS/mbedtls-framework/pull/85
- [ ] **3.6 PR**  not required because:  This is a major change that will not be backported.
- [ ] **2.28 PR** not required because:  This is a major change that will not be backported.
- **tests**  not required because: We are not introducing new features but the current tests are adjusted for the migration.

### CODE REVIEW

File: scripts/config.py [hunk 1]
```diff
@@ -419,10 +419,6 @@ def set(self, name, value=None):
             if name in PSA_UNSTABLE_FEATURE:
                 raise ValueError(f'Feature is unstable: \'{name}\'')
 
-            # The default value in the crypto config is '1'
-            if not value:
-                value = '1'
-
```
Comment (ronald-cron-arm) [inline scripts/config.py:425]: I think we should keep this for PSA_WANT symbols.

**Review by ronald-cron-arm — COMMENTED** [review]
I've had a first overall look. I've left a few comments. Otherwise, I think it would be beneficial to rebase on top of  #9771 now. You can then remove all MBEDTLS_PSA_CRYPTO_CONFIG. Also instead of moving MBEDTLS_USE_PSA_CRYPTO across config files, just remove it, it is not a configuration option anymore. Finally,  it would be nice to rework the changes around MBEDTLS_PSA_CRYPTO_USER/CONFIG_FILE now that the final state seems to be suitable.

File: tests/scripts/all-core.sh [hunk 2]
```diff
@@ -165,6 +165,8 @@ pre_load_components () {
 pre_initialize_variables () {
     if in_mbedtls_repo; then
         CONFIG_H='include/mbedtls/mbedtls_config.h'
+        CONFIG_TEST_DRIVER_H='tests/include/test/drivers/config_test_driver.h'
+        CONFIG_CRYPTO_TEST_DRIVER_H='tests/include/test/drivers/config_test_driver_tf_psa_crypto.h'
```
Comment (ronald-cron-arm) [inline tests/scripts/all-core.sh:169]:
```suggestion
        CRYPTO_CRYPTO_TEST_DRIVER_H='tests/include/test/drivers/crypto_config_test_driver.h'
```

**Review by ronald-cron-arm — CHANGES_REQUESTED** [review]
This looks quite good to me. I have a few minor comments. Otherwise, regarding the test_suite_config.psa_boolean errors reported by `analyze_outcomes.py`, it is just that the test cases have been moved from test_suite_config.mbedtls_boolean to test_suite_config.psa_boolean as expected and consequently they have to be moved in analyze_outcomes.py in IGNORED_TESTS from the test_suite_config.mbedtls_boolean key to the test_suite_config.psa_boolean one.

File: tests/Makefile [hunk 3]
```diff
@@ -344,12 +344,12 @@ libtestdriver1.a:
...
 	# Set the PSA cryptography configuration for the test library.
 	# The configuration is created by joining the base
-	    # ../tf-psa-crypto/tests/configs/crypto_config_test_driver.h,
-		# with the the library's PSA_WANT_* macros
-		# ./tf-psa-crypto/include/psa/crypto_config.h
-		# and then extended with entries of
-        # ../tf-psa-crypto/tests/configs/crypto_config_test_driver_extension.h
-        # to mirror the PSA_ACCEL_* macros.
+	# ../tf-psa-crypto/tests/configs/crypto_config_test_driver.h,
+	# with the the library's PSA_WANT_* macros
```
Comment (ronald-cron-arm) [inline tests/Makefile:348]:
```suggestion
	# with the the library's PSA_WANT_* macros extracted from
```

File: tests/scripts/components-build-system.sh [hunk 4]
```diff
@@ -184,15 +184,16 @@ component_build_cmake_custom_config_file () {
     cp tf-psa-crypto/include/psa/crypto_config.h $OUT_OF_SOURCE_DIR/full_crypto_config.h
     cd "$OUT_OF_SOURCE_DIR"
     echo '#error "cmake -DMBEDTLS_CONFIG_FILE is not working."' > "$MBEDTLS_ROOT_DIR/$CONFIG_H"
-    cmake -DGEN_FILES=OFF -DMBEDTLS_CONFIG_FILE=/$MBEDTLS_ROOT_DIR/$OUT_OF_SOURCE_DIR/full_config.h -DTF_PSA_CRYPTO_CONFIG_FILE=/$MBEDTLS_ROOT_DIR/$OUT_OF_SOURCE_DIR/full_crypto_config.h "$MBEDTLS_ROOT_DIR"
+    cmake -DGEN_FILES=OFF -DMBEDTLS_CONFIG_FILE=full_config.h -DTF_PSA_CRYPTO_CONFIG_FILE=full_crypto_config.h "$MBEDTLS_ROOT_DIR"
     make
 
-    msg "build: cmake with -DMBEDTLS_CONFIG_FILE + -DMBEDTLS_USER_CONFIG_FILE"
+    msg "build: cmake with -DMBEDTLS_CONFIG_FILE + -DMBEDTLS_USER_CONFIG_FILE + -DTF_PSA_CRYPTO_CONFIG_FILE"
```
Comment (ronald-cron-arm) [inline tests/scripts/components-build-system.sh:190]:
```suggestion
    msg "build: cmake with -DMBEDTLS/TF_PSA_CRYPTO_CONFIG_FILE + -DMBEDTLS/TF_PSA_CRYPTO_USER_CONFIG_FILE"
```

File: tests/scripts/components-build-system.sh [hunk 5]
```diff
@@ -216,15 +217,16 @@ component_build_cmake_custom_config_file () {
     cp tf-psa-crypto/include/psa/crypto_config.h full_crypto_config.h
 
     echo '#error "cmake -DMBEDTLS_CONFIG_FILE is not working."' > "$MBEDTLS_ROOT_DIR/$CONFIG_H"
-    cmake -DGEN_FILES=OFF -DTF_PSA_CRYPTO_CONFIG_FILE=/$MBEDTLS_ROOT_DIR/full_crypto_config.h -DMBEDTLS_CONFIG_FILE=/$MBEDTLS_ROOT_DIR/full_config.h .
+    cmake -DGEN_FILES=OFF -DTF_PSA_CRYPTO_CONFIG_FILE=full_crypto_config.h -DMBEDTLS_CONFIG_FILE=full_config.h .
     make
 
-    msg "build: cmake (in-tree) with -DMBEDTLS_CONFIG_FILE + -DMBEDTLS_USER_CONFIG_FILE"
+    msg "build: cmake (in-tree) with -DMBEDTLS_CONFIG_FILE + -DMBEDTLS_USER_CONFIG_FILE + -DTF_PSA_CRYPTO_CONFIG_FILE"
```
Comment (ronald-cron-arm) [inline tests/scripts/components-build-system.sh:223]:
```suggestion
    msg "build: cmake (in-tree) with -DMBEDTLS/TF_PSA_CRYPTO_CONFIG_FILE + -DMBEDTLS/TF_PSA_CRYPTO_USER_CONFIG_FILE"
```

File: tests/ssl-opt.sh [hunk 6]
```diff
@@ -235,7 +235,7 @@ get_options "$@"
 # and a comment after "#define SYMBOL"). The variable contains a
 # space-separated list of symbols.
 if [ "$LIST_TESTS" -eq 0 ];then
-    CONFIGS_ENABLED=" $(echo `$P_QUERY -l` )"
+    CONFIGS_ENABLED=" $(echo `$P_QUERY -l` ) "
```
Comment (ronald-cron-arm) [inline tests/ssl-opt.sh:238]: Please add a comment above about this.

File: tests/scripts/components-build-system.sh [hunk 7]
```diff
@@ -177,21 +178,26 @@ component_build_cmake_custom_config_file () {
...
     # that nothing else depends on).
-    echo '#undef MBEDTLS_NIST_KW_C' >user_config.h
+    echo '#undef MBEDTLS_SSL_ALL_ALERT_MESSAGES' >user_config.h
+    echo '#undef MBEDTLS_NIST_KW_C' >crypto_user_config.h
 
-    cmake -DGEN_FILES=OFF -DMBEDTLS_CONFIG_FILE=full_config.h -DMBEDTLS_USER_CONFIG_FILE=user_config.h "$MBEDTLS_ROOT_DIR"
+    cmake -DGEN_FILES=OFF -DMBEDTLS_CONFIG_FILE=full_config.h -DMBEDTLS_USER_CONFIG_FILE=user_config.h -DTF_PSA_CRYPTO_CONFIG_FILE=full_crypto_config.h -DTF_PSA_CRYPTO_USER_CONFIG_FILE=crypto_user_config.h "$MBEDTLS_ROOT_DIR"
     make
-    not programs/test/query_compile_time_config MBEDTLS_NIST_KW_C
+    not programs/test/query_compile_time_config MBEDTLS_SSL_ALL_ALERT_MESSAGES
```
Comment (ronald-cron-arm) [inline tests/scripts/components-build-system.sh:198]:
```suggestion
    not programs/test/query_compile_time_config MBEDTLS_SSL_ALL_ALERT_MESSAGES
    not programs/test/query_compile_time_config MBEDTLS_NIST_KW_C
```

File: tests/scripts/components-build-system.sh [hunk 8]
```diff
@@ -200,25 +206,29 @@ component_build_cmake_custom_config_file () {
...
     # that nothing else depends on).
-    echo '#undef MBEDTLS_NIST_KW_C' >user_config.h
+    echo '#undef MBEDTLS_SSL_ALL_ALERT_MESSAGES' >user_config.h
+    echo '#undef MBEDTLS_NIST_KW_C' >crypto_user_config.h
 
-    cmake -DGEN_FILES=OFF -DMBEDTLS_CONFIG_FILE=full_config.h -DMBEDTLS_USER_CONFIG_FILE=user_config.h .
+    cmake -DGEN_FILES=OFF -DMBEDTLS_CONFIG_FILE=full_config.h -DMBEDTLS_USER_CONFIG_FILE=user_config.h -DTF_PSA_CRYPTO_CONFIG_FILE=full_crypto_config.h -DTF_PSA_CRYPTO_USER_CONFIG_FILE=crypto_user_config.h .
     make
-    not programs/test/query_compile_time_config MBEDTLS_NIST_KW_C
+    not programs/test/query_compile_time_config MBEDTLS_SSL_ALL_ALERT_MESSAGES
```
Comment (ronald-cron-arm) [inline tests/scripts/components-build-system.sh:231]:
```suggestion
    not programs/test/query_compile_time_config MBEDTLS_SSL_ALL_ALERT_MESSAGES
    not programs/test/query_compile_time_config MBEDTLS_NIST_KW_C
```

File: configs/ext/tfm_mbedcrypto_config_profile_medium.h [hunk 9]
```diff
@@ -234,7 +234,7 @@
  * mechanism). You may still freely enable additional `MBEDTLS_XXX` symbols
  * in mbedtls_config.h.
  *
- * If the symbol #MBEDTLS_PSA_CRYPTO_CONFIG_FILE is defined, it specifies
+ * If the symbol #TF_PSA_CRYPTO_CONFIG_FILE is defined, it specifies
```
Comment (ronald-cron-arm) [inline configs/ext/tfm_mbedcrypto_config_profile_medium.h:237]: I forgot about it previously but we need to update configs/ext/README.md explaining the current situation where the TF-M config files are not directly from TF-M but have been adapted to the Mbed TLS configuration file split and more generally to the Mbed TLS repository split.

File: docs/proposed/config-split.md [hunk 10]
```diff
@@ -110,6 +110,7 @@ section dedicated to non boolean configuration options anymore.
 ### In `tf_psa_crypto_config.h`, we have:
 #### SECTION Platform abstraction layer
 ```
+//#define MBEDTLS_MEMORY_BUFFER_ALLOC_C
```
Comment (ronald-cron-arm) [inline docs/proposed/config-split.md:113]: Rather between MBEDTLS_MEMORY_BACKTRACE and MBEDTLS_MEMORY_DEBUG, alphabetic order.

File: include/mbedtls/mbedtls_config.h [hunk 11]
```diff
@@ -962,410 +302,38 @@
...
- *
- * Enable the prime-number generation code.
- *
- * Requires: MBEDTLS_BIGNUM_C
- */
-#define MBEDTLS_GENPRIME
-
-/**
- * \def MBEDTLS_FS_IO
+ * \def MBEDTLS_SELF_TEST
```
Comment (ronald-cron-arm) [inline include/mbedtls/mbedtls_config.h:305]: Should be moved to crypto_config.h section "General and test configuration options". Same for MBEDTLS_TEST_CONSTANT_FLOW_MEMSAN, MBEDTLS_TEST_CONSTANT_FLOW_VALGRIND and MBEDTLS_TEST_HOOKS below.

File: tf-psa-crypto/include/psa/crypto_config.h [hunk 12]
```diff
@@ -2707,6 +2707,35 @@
  */
 #define MBEDTLS_SHA3_C
 
+/**
+ * \def MBEDTLS_CIPHER_NULL_CIPHER
```
Comment (ronald-cron-arm) [inline tf-psa-crypto/include/psa/crypto_config.h:2711]: We need to update config-split.md accordingly.

**Review by ronald-cron-arm — CHANGES_REQUESTED** [review]
This looks almost good to me. See my final comments. Otherwise, config-split.md reorganizes sections in mbedtls_config.h and this is not done in this PR. I have also not checked the order of the options in crypto_config.h. This PR is already big enough thus I propose to do these in a follow-up PR.

File: tf-psa-crypto/include/psa/crypto_config.h [hunk 13]
```diff
@@ -147,4 +147,2595 @@
 #define PSA_WANT_KEY_TYPE_DH_KEY_PAIR_GENERATE  1
 //#define PSA_WANT_KEY_TYPE_DH_KEY_PAIR_DERIVE    1 /* Not supported */
```
Comment (gabor-mezei-arm) [inline tf-psa-crypto/include/psa/crypto_config.h:149]: Please add the section tags for the `PSA_WANT` macros.

```suggestion
/** \} name SECTION Cryptographic mechanism selection (PSA API) */

```

for Line 32:
```
/**
 * \name SECTION Cryptographic mechanism selection (PSA API)
 * 
 * \{
 */
```

**Review by gabor-mezei-arm — CHANGES_REQUESTED** [review]
Found only minor things otherwise looks good to me.

**Comment by minosgalanakis** [comment]
All comments have been addressed. I also cleaned up the commit history, and updated the framework pointer.

**Comment by ronald-cron-arm** [comment]
@minosgalanakis  Following the merge of #9781, we now have conflicts in `configs/ext/README.md` and `include/mbedtls/mbedtls_config.h`. For `configs/ext/README.md` I propose to just keep what has been done in #9781 thus to eventually not change anything in this PR. Otherwise please rebase to fix the conflicts in `include/mbedtls/mbedtls_config.h`.

File: ChangeLog.d/configuration-split.txt [hunk 14]
```diff
@@ -0,0 +1,8 @@
+Changes
+   * Configuration entries have been migrated from MBedTLS library
+     configuration file (include/mbedtls/mbedtls_config.h)
+     to the tf-psa-crypto one (tf-psa-crypto/include/psa/crypto_config.h)
+     as per design proposal (PR #9236). The test suites, as the built-in
+     and external (TF-M) configurations have been updated accordingly.
+     Configurations using `MBEDTLS_PSA_CRYPTO_CONFIG_FILE` as a pre-processor
+     directive, will have to specify the file's location using the build-system.
```
Comment (ronald-cron-arm) [inline ChangeLog.d/configuration-split.txt:8]: The last sentence seems too specific to me. I'd go for something more general like:
```suggestion
     To migrate custom Mbed TLS configurations where
     MBEDTLS_PSA_CRYPTO_CONFIG is disabled, you should first adapt them
     to the PSA configuration scheme based on PSA_WANT_XXX symbols
     (see psa-conditional-inclusion-c.md for more information).
     To migrate custom Mbed TLS configurations where
     MBEDTLS_PSA_CRYPTO_CONFIG is enabled, you should migrate the
     cryptographic and platform configuration options from mbedtls_config.h
     to crypto_config.h (see config-split.md for more information and configs/
     for examples).
```

**Review by ronald-cron-arm — CHANGES_REQUESTED** [review]
Some suggestions for the change log.

**Review by ronald-cron-arm — APPROVED** [review]
All my comments have been addressed, the rebase looks good to me thus approving.

**Comment by ronald-cron-arm** [comment]
@gabor-mezei-arm please have another look.

**Review by gabor-mezei-arm — APPROVED** [review]
LGTM

**Review by ronald-cron-arm — DISMISSED** [review]
Last update of analyze_outcomes.py and framework pointer look good to me.

**Comment by ronald-cron-arm** [comment]
@gabor-mezei-arm thanks for having another look.

**Review by gabor-mezei-arm — DISMISSED** [review]
LGTM

**Review by ronald-cron-arm — APPROVED** [review]
LGTM

**Review by adeaarm — COMMENTED** [review]
LGTM (I know, already merged :) )

**Pull request merged**